### PR TITLE
AWS S3 Url Reader 

### LINKED ISSUE #6250: Add AWS s3 integration for reading data from s3 buckets
## Feature Suggestion

I want to add AWS S3 integration to read data from a s3 bucket. This can be like the data reading integration that already exists for Google GCS ([GoogleGcsUrlReader.ts](https://github.com/backstage/backstage/blob/4045eaab952fec349e443990330bd1ac5b2c9c18/packages/backend-common/src/reading/GoogleGcsUrlReader.ts)).

I would like to work on this. 

## Context

For the purpose of reading .yaml files from an aws s3 bucket for service catalog data.

### PR DESCRIPTION
**Pull request by kjost-splunk**
## Hey, I just made a Pull Request!

Add AWS S3 URL reader to read data from a S3 bucket. 

Close #6250 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: .changeset/beige-trainers-run.md
Comment (OrkoHunter) [inline .changeset/beige-trainers-run.md:3]: These should be patch changes. Since we are still in 0.x releases, minor is only for breaking changes, and patch for new features!

File: packages/integration/src/awsS3/config.ts
Comment (OrkoHunter) [inline packages/integration/src/awsS3/config.ts:33]: Let's update the config schema as well! https://github.com/backstage/backstage/blob/master/packages/integration/config.d.ts

File: packages/integration/src/awsS3/config.ts
Comment (OrkoHunter) [inline packages/integration/src/awsS3/config.ts:56]: I think this should also work :D

```ts
  return { accessKeyId, secretAccessKey };
```

File: packages/backend-common/src/reading/AwsS3UrlReader.ts
Comment (OrkoHunter) [inline packages/backend-common/src/reading/AwsS3UrlReader.ts:41]: A comment describing why's for these two lines will help a lot! 🙏

File: packages/backend-common/src/reading/AwsS3UrlReader.ts
Comment (OrkoHunter) [inline packages/backend-common/src/reading/AwsS3UrlReader.ts:68]: > `integrations.awsS3` not found in app config. AWS S3 integration will use default AWS credentials if set in environment.


Perhaps this is a bit more explanatory in the logs, as this is printed when the backend starts. I'm also questioning if it should be `logger.debug` since a huge majority of people won't have this configured, and they won't even care, to be honest.

File: packages/backend-common/src/reading/AwsS3UrlReader.ts
Comment (OrkoHunter) [inline packages/backend-common/src/reading/AwsS3UrlReader.ts:104]: I think `read` is being deprecated in favor of `readUrl` which has broader request and response options. ETag based caching is possible only in `readUrl`.

If you look at GitHub URL Reader, you can see how `read` is just re-using `readUrl` as of now. I'd recommend implementing `readUrl` here.

**Review by OrkoHunter — COMMENTED** [review]
Awesome! Few suggestions here and there

File: docs/integrations/aws-s3/locations.md
Comment (andrewthauer) [inline docs/integrations/aws-s3/locations.md:23]: It feels like this should allow for a list of host / credential configs rather then a single one. This would make it possible to use different credentials for per host as well as be more consistent with existing integrations.

Comment (freben) [reply docs/integrations/aws-s3/locations.md:23]: Hmmm agreed, but i wonder what config patterns are best to use then.

https://docs.aws.amazon.com/AmazonS3/latest/userguide/access-bucket-intro.html

There are several possible url patterns that kinda overlap.

Will we need host + optional bucket as config params? Or making even host optional, implying the hosted amazonaws one in general if left out?

Comment (andrewthauer) [reply docs/integrations/aws-s3/locations.md:23]: That's true. There are a few more considerations then other URL readers in this regard. Using a public http url would likely require the bucket name as per this. It might also be nice to support the s3 protocol form as well at some point (e.g. `s3://some-bucket`). In this case, the bucket alone would need to be the indicator for which credentials to use.

Comment (freben) [reply docs/integrations/aws-s3/locations.md:23]: Yeah.

Maybe a good start is to check what parameters (including naming) that the node library accepts and consider mimicking

Comment (Xantier) [reply docs/integrations/aws-s3/locations.md:23]: Constructing the URL dynamically with strict validation would exclude many AWS-S3-like services from the mix as well. These alternatives use the same API as S3 so a small modification here would enable those as well with this same PR (with the addition of forcePathStyle config option). I'd recommend using a similar structure as in `plugins/techdocs-backend/config.d.ts`

Comment (freben) [reply docs/integrations/aws-s3/locations.md:23]: Ahh of course, there's prior art - thanks for reminding me, yes let's standardize how we express this.

File: .changeset/beige-trainers-run.md
Comment (kjost-splunk) [reply .changeset/beige-trainers-run.md]: Changed this to a patch!

File: packages/integration/src/awsS3/config.ts
Comment (kjost-splunk) [reply packages/integration/src/awsS3/config.ts]: Updated config schema!

File: packages/integration/src/awsS3/config.ts
Comment (kjost-splunk) [reply packages/integration/src/awsS3/config.ts]: Yes, that worked. Changed!

File: packages/backend-common/src/reading/AwsS3UrlReader.ts
Comment (kjost-splunk) [reply packages/backend-common/src/reading/AwsS3UrlReader.ts]: comments added!

File: packages/backend-common/src/reading/AwsS3UrlReader.ts
Comment (kjost-splunk) [reply packages/backend-common/src/reading/AwsS3UrlReader.ts]: Good point. Implemented readUrl

File: packages/backend-common/src/reading/AwsS3UrlReader.ts
Comment (kjost-splunk) [reply packages/backend-common/src/reading/AwsS3UrlReader.ts]: Changed the message to the one you suggested and changed it to `logger.debug`

File: docs/integrations/aws-s3/locations.md
Comment (kjost-splunk) [reply docs/integrations/aws-s3/locations.md]: We made the AWS integration support a list of host / credential configs. To do this, we followed what had been done for the other URL readers that support a list like adding the `AwsS3Integrations` class that implements `ScmIntegration` . The other Url readers that are included in `ScmIntegrations` seem to be associated with git repositories while AWS S3 is not. Is this a proper use of `ScmIntegrations`?

File: packages/integration/config.d.ts
Comment (benjdlambert) [inline packages/integration/config.d.ts:180]: I'm wondering if we need to make this a little more abstract here to support different forms of authentication with AWS? What about roles? would they be included here too?

Comment (benjdlambert) [reply packages/integration/config.d.ts:180]: Think this should also be `accessKeyId`

Comment (seant-splunk) [reply packages/integration/config.d.ts]: Hello, it looks like something similar is being done in the [Techdocs Backend](https://github.com/backstage/backstage/blob/master/plugins/techdocs-backend/config.d.ts) where `roleArn` is included alongside `accessKeyId` and `secretAccessKey`. Is this the direction we should consider for our use case?

Comment (Xantier) [reply packages/integration/config.d.ts]: I think it could be a good approach to follow the logic within `AwsS3Publish.buildCredentials` to generate the credentials object. That uses these two access & secret keys if they exist, defaults to using standard credentials provider chain if they dont, and finally optionally wraps the a role assumption around those credentials is roleArn is defined. 

For AWS stuff it might be useful to generalize this pattern IMO.

Comment (seant-splunk) [reply packages/integration/config.d.ts]: We followed the logic for `AwsS3Publish.buildCredentials` and incorporated `roleArn` in our configurations.

File: packages/integration/config.d.ts
Comment (freben) [inline packages/integration/config.d.ts:175]: accidental extra line, delete

File: packages/backend-common/src/reading/AwsS3UrlReader.ts
Comment (freben) [inline packages/backend-common/src/reading/AwsS3UrlReader.ts:36]:
```suggestion
   * Removes the leading '/' from the pathname to be processed
```

File: docs/integrations/aws-s3/locations.md
Comment (freben) [reply docs/integrations/aws-s3/locations.md]: I think you are doing the right thing here. This will open it up for usage in the URL readers automatically etc. I think the SCM part may have been a bit too strict naming on our part.

File: packages/backend-common/src/reading/AwsS3UrlReader.ts
Comment (freben) [inline packages/backend-common/src/reading/AwsS3UrlReader.ts:46]: So in practice this only supports hosted amazonaws.com S3. I think there were some discussions about wanting to support other S3 compatible solutions, of which there are several. But I'd be open for getting this shipped first, and then opening up for properly handling both

(1) the host config parameter, and
(2) URLs that are not on the form where the bucket is in the host name, but rather is the first path segment instead

**Review by freben — COMMENTED** [review]
This is looking really good! Some smaller things of note, but getting close.

Sorry that this has been taking some time because of vacations and stuff, let's get this shipped!

**Comment by seant-splunk** [comment]
Hello! I've made the suggested changes, but left the regex host checking in the AwsS3UrlReader for the moment, I am perhaps less aware of what URL's of other S3 compatible solutions are, so maybe more discussion is needed there?

**Comment by freben** [comment]
Alright, doing another review round then.

This is what I was referring to: https://github.com/backstage/backstage/blob/0c58dd73d0e98b7c08b5f770bbb30e3c1923e55f/plugins/techdocs-backend/config.d.ts?plain=1#L123

This is where it's read and then given to the S3 client: https://github.com/backstage/backstage/blob/0c58dd73d0e98b7c08b5f770bbb30e3c1923e55f/packages/techdocs-common/src/stages/publish/awsS3.ts?plain=1#L92

**Comment by freben** [comment]
@seant-splunk @kjost-splunk Awesome, this is almost done! I see two things still:

1. The extra comment line is still there.
2. The yarn.lock file doesn't seem to have been updates since you downgraded the dep version - you may need to run yarn install again and see that the extra block in yarn.lock disappears.

**Comment by freben** [comment]
After those two I'm super happy to get this merged!

**Comment by seant-splunk** [comment]
Ahh, man, so close! I've double checked that the extra comment has been removed and the `aws-sdk^2.952.0` block has been removed from the yarn.lock file

**Comment by freben** [comment]
Ugh, gosh darn it, through no fault of yours master has an issue. Fixing in https://github.com/backstage/backstage/pull/6793 - you can rebase on top of that already to be proactive, if you are daring. That way you won't have to wait for it to be merged and _then_ rebase on master and _then_ wait for us to merge yours :)

**Comment by seant-splunk** [comment]
Looks like prettier is complaining in the checks, is that something on our side we can solve?

**Comment by freben** [comment]
Ah yes, run `yarn prettier --write mkdocs.yml` to have it auto-fixed

**Comment by freben** [comment]
Oh no, now the test!

**Comment by freben** [comment]
Terribly sorry to trouble you again - you can unfortunately not just remove and rebuild the lock file. What you have to do is just to get the unmodified master version of the file and then run yarn install. It should ultimately contain only changes that are exactly what is relevant to your own modifications to dependencies - nothing else

**Comment by freben** [comment]
If you have the possibility of unlocking push rights to this fork, I can help fix up the last bit tomorrow.

**Comment by seant-splunk** [comment]
Ahh, so sorry to cause more little issues! I've taken the `yarn.lock` from master and ran `yarn install`. If this fails, I'll look into unlocking push rights

**Review by freben — APPROVED** [review]
Let's gooooooo

**Pull request merged**
